### ISSUE TITLE
void inference lost using a specific parameter syntax while implementing an interface, causes strict mode errors

### ISSUE THREAD

**Issue by jeremiegirault** [Microsoft/TypeScript#30087]
Hi there & thanks for the good work. I've read a few pages of issues and could not find a relevant so I'm posting one hoping that it won't be a duplicate.

I found a strange behavior around `void` type inference in generics in strict mode and managed to isolate a reproducing sample. I wanted to contribute it to you if it can be of any help.

**TypeScript Version:**  3.4.0-dev.20190223

**Search Terms:** void, generic, strict

**Code**

```ts
interface Startable<StartOptions> {
  start (options: StartOptions): void
}

type StartableOptions<T> = T extends Startable<infer U> ? U : never

class X1 implements Startable<number> {
  public start (num: number) {
    console.log(num)
  }
}

class X2 implements Startable<void> {
  public start (v: void) {
    console.log('X2', v)
  }
}

class X3 implements Startable<void> {
  public start () { // as expected: no typecheck error on X3
    console.log('X3')
  }
}

type OX1 = StartableOptions<X1> // OX1 = number
type OX2 = StartableOptions<X2> // OX2 = void
type OX3 = StartableOptions<X3> // OX3 = {} => expected to be void !

const x1: OX1 = 42
const x2: OX2 = undefined
const x3: OX3 = undefined // Type 'undefined' is not assignable to type '{}'.
```

**Expected behavior:**

Type of `OX3` is expected to be `void`.
reason:
 - it typechecks the `Startable<void>` implementation !
 - OX2 properly infers `void` using the explicit `void` syntax.

**Actual behavior:**

Type of `OX3` resolves to `{}` and the code fails to compile.

```
tsc test.ts --strict
test.ts:31:7 - error TS2322: Type 'undefined' is not assignable to type '{}'.

31 const x3: OX3 = undefined
         ~~

Found 1 error.

error Command failed with exit code 2.
```

**Playground Link:** 

Enable `strictNullChecks` by clicking (it is not recorded in url)
[here](https://www.typescriptlang.org/play/#src=interface%20Startable%3CStartOptions%3E%20%7B%0D%0A%20%20start%20(options%3A%20StartOptions)%3A%20void%0D%0A%7D%0D%0A%0D%0Atype%20StartableOptions%3CT%3E%20%3D%20T%20extends%20Startable%3Cinfer%20U%3E%20%3F%20U%20%3A%20never%0D%0A%0D%0Aclass%20X1%20implements%20Startable%3Cnumber%3E%20%7B%0D%0A%20%20public%20start%20(num%3A%20number)%20%7B%0D%0A%20%20%20%20console.log(num)%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aclass%20X2%20implements%20Startable%3Cvoid%3E%20%7B%0D%0A%20%20public%20start%20(v%3A%20void)%20%7B%0D%0A%20%20%20%20console.log('X2'%2C%20v)%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aclass%20X3%20implements%20Startable%3Cvoid%3E%20%7B%0D%0A%20%20public%20start%20()%20%7B%20%2F%2F%20as%20expected%3A%20no%20typecheck%20error%20on%20X3%0D%0A%20%20%20%20console.log('X3')%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Atype%20OX1%20%3D%20StartableOptions%3CX1%3E%20%2F%2F%20OX1%20%3D%20number%0D%0Atype%20OX2%20%3D%20StartableOptions%3CX2%3E%20%2F%2F%20OX2%20%3D%20void%0D%0Atype%20OX3%20%3D%20StartableOptions%3CX3%3E%20%2F%2F%20OX3%20%3D%20%7B%7D%20%3D%3E%20expected%20to%20be%20void%20!%0D%0A%0D%0Aconst%20x1%3A%20OX1%20%3D%2042%0D%0Aconst%20x2%3A%20OX2%20%3D%20undefined%0D%0Aconst%20x3%3A%20OX3%20%3D%20undefined%0D%0A)

**Related Issues:**
Not found. Searched open and closed issues

**Comment by RyanCavanaugh** [comment]
I'm inferring that you believe `void` is the type of a parameter that doesn't exist, but that is not the case. `void` should really only be used in the return type of functions; putting it in a parameter position doesn't make much sense.

There are other ways to write this to get the behavior you want (basically you need a second path in the conditional).
